src/utils/communication_buffer.hpp

```diff
@@ -0,0 +1,355 @@
+//========================================================================================
+// (C) (or copyright) 2020-2022. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+
+#ifndef UTILS_COMMUNICATION_BUFFER_HPP_
+#define UTILS_COMMUNICATION_BUFFER_HPP_
+
+#include <iostream>
+#include <memory>
+#include <type_traits>
+#include <unordered_map>
+#include <utility>
+
+#ifdef MPI_PARALLEL
+#include <mpi.h>
+
+#define request_t MPI_Request
+#define comm_t MPI_Comm
+#else
+#define request_t int
+#define comm_t int
+#endif
+
+namespace parthenon {
+namespace impl {
+
+#ifdef MPI_PARALLEL
+// MPIType<A> -> MPIType<A, bool_t<true>> and then
+// specialization checks are performed. This is why the
+// default parameter is required.
+template <bool>
+struct bool_t {};
+
+template <class T, class = bool_t<true>>
+struct MPIType;
+
+template <class U, class T>
+using MPI_type_check = std::is_same<U, std::remove_pointer_t<T>>;
+
+template <class T>
+struct MPIType<T, bool_t<MPI_type_check<double, T>::value>> {
+  MPI_Datatype static value() noexcept { return MPI_DOUBLE; }
+};
+
+template <class T>
+struct MPIType<T, bool_t<MPI_type_check<int, T>::value>> {
+  MPI_Datatype static value() noexcept { return MPI_INT; }
+};
+
+template <class T>
+struct MPIType<T, bool_t<MPI_type_check<bool, T>::value>> {
+  MPI_Datatype static value() noexcept { return MPI_CXX_BOOL; }
+};
+#endif
+} // namespace impl
+
+using namespace impl;
+
+//             Read    Write
+//    stale:             X
+// sending*:
+// received:     X
+enum class BufferState { stale, sending, sending_null, received, received_null };
+
+enum class BuffCommType { sender, receiver, both };
+
+template <class T>
+class CommBuffer {
+  // Need specializations to be friends with each other
+  template <typename U>
+  friend class CommBuffer;
+
+  std::shared_ptr<BufferState> state_;
+  std::shared_ptr<BuffCommType> comm_type_;
+  std::shared_ptr<bool> recv_start_called_;
+  std::shared_ptr<request_t> my_request_;
+
+  int my_rank;
+  int tag_;
+  int send_rank_;
+  int recv_rank_;
+  comm_t comm_;
+
+  using buf_base_t = std::remove_pointer_t<decltype(std::declval<T>().data())>;
+  buf_base_t null_buf_;
```

Comment:
So I don't fully understand the null message machinery. Do you leave `null_buf_` uninitialized and send a message of size zero? Is that unique? I.e., this will never happen by accident?

Maybe we should explicitly initialize the null buffer with a single signal NaN, even if we don't send it?

Also if you send a message of size zero what does MPI actually do? I does it send metadata?

Reply:
Yes, I leave `null_buf_` initialized and send a message of size zero. All messages are first received using `MPI_IProbe` to check their size and then either read into the actual buffer or are read into the null buffer (which shouldn't do anything besides make the message show as read). I believe MPI sends all of the metadata for a message of size zero; I don't think the code would work if it did not. 

It is a good point that you could send a message of size zero if the input buffer has size zero. I have added a `PARTHENON_REQUIRE` to check for that. I think there is no use case for sending buffers that are of size zero. 

I also added a default initialization to `signaling_NaN`. I don't think this should matter, but in case it ever becomes an issue this should at least help to make it less quiet.

Reply:
Sounds good. My thinking about the sgnaling NaN was in case somehow someone accidentally points to the null buffer, a deliberate choice inside is probably better than whatever random bits you end up looking at.